languages/thingtalk/load-thingpedia.js

```diff
@@ -834,6 +837,86 @@ class ThingpediaLoader {
         ));
     }
 
+    _loadDate() {
+        const months = [
+            ['january', 'jan'],
+            ['february', 'feb'],
+            ['march', 'mar'],
+            ['april', 'apr'],
+            ['may'],
+            ['june', 'jun'],
+            ['july', 'jul'],
+            ['august', 'aug'],
+            ['september', 'sep', 'sept'],
+            ['october', 'oct'],
+            ['november', 'nov'],
+            ['december', 'dec']
+        ];
+        const weekdays = {
+            'monday': ['monday', 'mon'],
+            'tuesday': ['tuesday', 'tue'],
+            'wednesday': ['wednesday', 'wed'],
+            'thursday': ['thursday', 'thu'],
+            'friday': ['friday', 'fri'],
+            'saturday': ['saturday', 'sat'],
+            'sunday': ['sunday', 'sun'],
+        };
+
+        let expansion;
+        this._grammar.declareSymbol('constant_date_range');
+        this._grammar.declareSymbol('constant_Date');
+
+        // year
+        expansion = [new this._runtime.NonTerminal('constant_Number')];
+        this._grammar.addRule('constant_date_range', expansion, this._runtime.simpleCombine((value) => [
+            makeDate(new Ast.DatePiece(value.value, null, null, null), '+', null),
+            makeDate(new Ast.DatePiece(value.value, null, null, null), '+', new Ast.Value.Measure(1, 'year'))
+        ]));
+        expansion = [new this._runtime.NonTerminal('constant_Number'), 's']; // 90s
+        this._grammar.addRule('constant_date_range', expansion, this._runtime.simpleCombine((value, _2) => [
+            makeDate(new Ast.DatePiece(value.value, null, null, null), '+', null),
+            makeDate(new Ast.DatePiece(value.value, null, null, null), '+', new Ast.Value.Measure(10, 'year'))
+        ]));
+
+        // year + month
+        for (let month = 1; month <= 12; month ++) {
+            for (let phrase of months[month-1]) {
+                expansion = [phrase, 'of', new this._runtime.NonTerminal('constant_Number')];
```

Comment:
If we care about translation (and I think we do), we really cannot do this loop stuff. Why do you even need it anyway? Why are month and day not just a plain non-terminal?

Reply:
this is already moved to `constants.genie`, loops just makes it slightly easier to write...

Reply:
Easier to write, maybe. But not easier to read, and that's more important. It would be a lot more readable with a non-terminal for "day", a non-terminal for "month", and proper grammar to combine the two. Plus it won't work with my translation PR to have a loop.

Reply:
https://github.com/stanford-oval/genie-toolkit/blob/wip/cikm-camera-ready/languages/thingtalk/en/constants.genie#L146
this is what i have right now. why this cannot be translated? is it because of `#(phrase)`?

Reply:
Yes. It makes the templates dynamic, so we cannot statically extract all the templates and prepare a translation file. (You should take a look at #285 which I just pushed.

Reply:
got it. I will fix this.

Reply:
fixed in 2b3515b